File: echidna_client.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <cstring>
#include <zmq.hpp>
#include <Eigen/Dense>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/core/eigen.hpp>
#include "getoptpp/getopt_pp.h"


using namespace zmq;
using namespace std;
using namespace Eigen;
using namespace cv;
using namespace GetOpt;

bool bDisplay = false;
unsigned int NUM_SEC = 5;
unsigned int HI_CHAN = 7;
unsigned int LO_CHAN = 0;
unsigned int NUM_SAMPLES = 1000;
unsigned int SAMP_RATE = 60000;
float MAX_VOLT_F = 5.0f;
string server_connection_str("localhost");
int server_port = 5555;

static void show_usage(char *progname);

int main(int argc, char* argv[]) 
{

    GetOpt_pp ops(argc, argv);

    if (ops >> OptionPresent('h',"help")) {
        show_usage(argv[0]);
        return 0;
    }
    
    if (ops >> OptionPresent('D',"display")) {
        bDisplay = true;
    }

    ops >> Option('H',"high_channel",HI_CHAN);
    if(HI_CHAN > 7) {
        show_usage(argv[0]);
        return 1;
    }

    ops >> Option('L',"low_channel",LO_CHAN);
    if(LO_CHAN > 7) {
        show_usage(argv[0]);
        return 1;
    }
    ops >> Option('s',"samples",NUM_SAMPLES);
    if(NUM_SAMPLES>10000){
        show_usage(argv[0]);
        return 1;
    }    

    if(ops >> Option('S',"server",server_connection_str)) {
        cout << "Connect to server: " << server_connection_str << endl;
    }

   if(ops >> Option('P',"port",server_port)) {
        cout << "Connect to port: " << server_port << endl;
   } 

   stringstream server_connection;
   server_connection << "tcp://" << server_connection_str << ":" << server_port;

   cout << "Connect to server at: " << server_connection.str()<<endl;        

    context_t context(1);
    socket_t socket(context, ZMQ_SUB);
    socket.setsockopt(ZMQ_SUBSCRIBE, "", 0);

    cout << "Connecting to Echidna server" << endl;
    
    Mat image, big_image;
    Size sz(NUM_SAMPLES,50*(HI_CHAN-LO_CHAN+1));  
        
    if(bDisplay) {
        cout << "Show Display" << endl;
    }
    namedWindow("Client Display window", CV_WINDOW_FULLSCREEN); 
    socket.connect("tcp://localhost:5555");
    long count = 0;
    for(;;) {
        float *pf;
        message_t data_msg;
        socket.recv(&data_msg);
        pf = (float*) data_msg.data(); 
        Map<MatrixXf> mf(pf,(HI_CHAN-LO_CHAN+1),NUM_SAMPLES);
        cout << " Received: " << count << endl;
        if (bDisplay) {
            eigen2cv((MatrixXf)mf, image);
            big_image = big_image.zeros(sz,image.type());
            resize(image, big_image, sz);
            //cout << "Size: " << big_image.size().height << ","<< big_image.size().width << endl;
            imshow("Client Display window", big_image);
        }
        count++;
        if(waitKey(1)>=0) {
                break;
        }
    }
    return 0;
}

void show_usage(char* progname) 
{
    cout << "Usage: " << progname << " [OPTION]" << endl;
    cout << "Read data from the USB-1608FS Data Acquisition\n"
            "Use the sound card to generate a sine wave\n"
            "-H, --high_channel\t\t"
            "select the highest channel on the USB DAQ to sample (0-7)\n"            
            "-L, --low_channel\t\t"
            "select the lowest channel o nthe USB DAQ to sample (0-7)\n"
            "-s, --sample\t\t"
            "choose the number of samples to per image\n" << endl;
}

File: echidna_client.py
import sys
import os
import argparse
import logging
import datetime
import zmq
import struct
import numpy as np

logger = logging.getLogger("Echidna Client")
hdlr = logging.StreamHandler(sys.stdout)
hdlr.setLevel(logging.DEBUG)
logger.addHandler(hdlr)
logger.setLevel(logging.DEBUG)



parser = argparse.ArgumentParser(description="Echidna client")

parser.add_argument('--server', '-s', type=str, default = "localhost",
                    help="select the server to connect to")
parser.add_argument('--port','-p',type=int, default=5555,
                    help="select the port to connect to")


opts = parser.parse_args()

server_connect_str = "tcp://%s:%d" % (opts.server,opts.port)

print "Connect: %s" % (server_connect_str)

def main():
    
    context = zmq.Context()
    socket = context.socket(zmq.SUB)
    socket.setsockopt(zmq.SUBSCRIBE, '')

    socket.connect(server_connect_str)

    count = 0;
    while True:
        msg = socket.recv_multipart()
        print "Multiparts: %d" % len(msg)
        num_channels = struct.unpack("<I",msg[0])[0]
        print "Number of channels %d" % num_channels
        num_samples = struct.unpack("<I",msg[1])
        print "Number of Samples %d" % num_samples[0]
        print "Count: %d | Length: %d " % (count, len(msg[2]))
        data = np.reshape(np.frombuffer(msg[2], dtype=np.float64),(-1,num_channels))

        print data[0:2,:]
        count = count + 1


if __name__ == "__main__":
    data = main()


